### PR TITLE
Enabling group-specific admin permissions for leaves and leave artifacts

### PR DESCRIPTION
**Pull request by jxjj**
This permits group managers (members with `admin` flag) to view/modify leaves and leave artifacts for fellow group members.

## Background

Before, anyone with a `group admin` role or `view leaves` permission could see and edit leave details for anyone across the app. The `admin` flag on a group member didn't impact leave permissions. 

 Now, if you're a manager of your group, you can manage leave info for fellow group members, and not members of other groups where you aren't an admin.
 
This means fewer people should need to be global group admins.

## How it works

We're using [Laravel Policies](https://laravel.com/docs/11.x/authorization#creating-policies) to control access to models. The policy serves as a source of truth for what actions a user is permitted to take.

Policies work great for the API, but communicating policies to the frontend poses some questions: When you land on a page for User X, does the current user have permissions to view all the leaves? create a new one? edit/delete an existing one?

For individual resource items, we're adopted the approach [discussed on the laracasts forum](https://laracasts.com/discuss/channels/vue/laravel-policies-and-gates-in-vue): a resource (leave) will have permission info appended to the resource:

```json5
{ 
   // ...normal leave stuff...
   currentUserCan: {
      update: true,
      delete: true,
   }
}
```

For general resource actions like `viewAny` or `create`,  appending won't work (there may not be any items to append the permissions to). So, we add API endpoints for checking permissions. For example, `/api/permissions/users/:leaveOwner/leaves`:

```json5
{
  // current user can view any leaves owned by `leaveOwner`
  viewAny: true,

  // but they cannot create new leaves owned by `leaveOwner`
  create: false,
}
```

This introduces an async permissions check to the front-end. It's currently, uncached... but we could add that if we're seeing a lot of requests or need to dedupe.

## Other changes

- We created constants for `PermissionRoles` and `Permissions` to avoid mistakes and make it easier to manage.
- renamed `group admin` role to `global group admin` to differentiate it from group members with `admin` flag.
- added some factories for testing, like `MembershipFactory` for creating group members.
- on the frontend, permissions are now managed by a separate PermissionStore.
- expanded and updated tests to check everything works as expected

### CODE REVIEW

**Comment by jxjj** [comment]
updated policies to account for users who may be instructors but technically members of a group. 

Todo:
- write a test to convince myself that's it's actually working
- fix existing tests. Now that the policy checks may involve an api call to bandaid, I may need to move the broken api tests to use pest so that we can mock Bandaid.

I didn't really see any huge performance hits, but let me know if we need to optimize anything. One easy thing would be to hang onto the getDeptInstructors cache longer.

File: app/Library/UserService.php
Comment (cmcfadden) [inline app/Library/UserService.php:173]: could I convince you to change this to `contains`? `some` is an alias and it never makes sense to me to read it like that

File: app/Policies/LeaveArtifactPolicy.php
Comment (cmcfadden) [inline app/Policies/LeaveArtifactPolicy.php:65]: wanna move these checks to a method (can just be a private controller method) so the two methods here call the same logic?

Comment (cmcfadden) [reply app/Policies/LeaveArtifactPolicy.php:65]: or all the methods I guess, I hadn't scrolled down

**Comment by jxjj** [comment]
Updates
- tests fixed. any api tests were migrated to PestPHP. This was necessary since policies now may do a call to Bandaid to check if the user is an instructor in a department.
- added test for granting managers permissions to view/edit user leaves when the user is not an official member of a group they manage, but is an instructor in a department they manage
- `some` changed to `contains`
- simplified policy logic. Most Leave and LeaveArtifact polices now defer to `viewAnyUserLeaves` or `modifyUserLeaves` methods

**Review by cmcfadden — APPROVED** [review]
looks good. We'll get it on staging and exercise it with some real world users. Can tap the improving process team.

**Comment by jxjj** [comment]
noticed a small bug where cancelling creation of a new leave doesn't remove it from the front-end store. Will fix that before merging.

**Comment by jxjj** [comment]
Looks like the bug is in prod already. I'll put the fix in a separate PR.

**Pull request merged**
